src/parser.rs

```diff
@@ -551,175 +557,199 @@ pub enum SegmentTag {
+        ),
+        map(tag("#EXT-X-DISCONTINUITY"), |_| SegmentTag::Discontinuity),
+        map(pair(tag("#EXT-X-KEY:"), key), |(_, key)| {
+            SegmentTag::Key(key)
+        }),
+        map(pair(tag("#EXT-X-MAP:"), extmap), |(_, map)| {
+            SegmentTag::Map(map)
+        }),
+        map(
+            pair(tag("#EXT-X-PROGRAM-DATE-TIME:"), consume_line),
+            |(_, line)| SegmentTag::ProgramDateTime(line),
+        ),
+        map(
+            pair(tag("#EXT-X-DATE-RANGE:"), consume_line),
+            |(_, line)| SegmentTag::DateRange(line),
+        ),
+        map(ext_tag, SegmentTag::Unknown),
+        map(comment_tag, SegmentTag::Comment),
+        map(consume_line, SegmentTag::Uri),
+    ))(i)
+}
+
+pub fn duration_title_tag(i: &[u8]) -> IResult<&[u8], (f32, Option<String>)> {
+    map(
+        tuple((
+            float,
+            opt(char(',')),
+            opt(map_res(is_not("\r\n,"), from_utf8_slice)),
+            take(1usize),
+            opt(char(',')),
+        )),
+        |(duration, _, title, _, _)| (duration, title),
+    )(i)
+}
+
+pub fn key(i: &[u8]) -> IResult<&[u8], Key> {
+    map(key_value_pairs, Key::from_hashmap)(i)
+}
+
+pub fn extmap(i: &[u8]) -> IResult<&[u8], Map> {
+    map_res(key_value_pairs, |attrs| -> Result<Map, &str> {
+        let uri = attrs.get("URI").cloned().unwrap_or_default();
+        let byte_range = attrs
+            .get("BYTERANGE")
+            .map(|range| match byte_range_val(range.as_bytes()) {
+                IResult::Ok((_, range)) => Ok(range),
+                IResult::Err(_) => Err("invalid byte range"),
+            })
+            .transpose()?;
+
+        Ok(Map { uri, byte_range })
+    })(i)
+}
 
 // -----------------------------------------------------------------------------------------------
 // Basic tags
 // -----------------------------------------------------------------------------------------------
 
-named!(pub m3u_tag<String>,
-     map_res!(tag!("#EXTM3U"), from_utf8_slice)
-);
-
-named!(pub version_tag<usize>,
-    do_parse!(
-        tag!("#EXT-X-VERSION:") >> version: map_res!(digit1, str::from_utf8) >>
-        (version.parse().unwrap_or_default())
-    )
-);
-
-named!(pub start_tag<Start>,
-    do_parse!(tag!("#EXT-X-START:") >> attributes:key_value_pairs >>
-        (Start::from_hashmap(attributes))
-     )
-);
-
-named!(pub ext_tag<ExtTag>,
-    do_parse!(
-        tag!("#EXT-")
-        >> tag: map_res!(is_not!("\r\n:"), from_utf8_slice)
-        >> opt!(tag!(":"))
-        >> rest: opt!(map_res!(is_not!("\r\n"), from_utf8_slice))
-        >> take!(1)
-        >> (
-            ExtTag { tag, rest }
-        )
-    )
-);
-
-named!(pub comment_tag<String>,
-    do_parse!(
-        tag!("#") >> text: map_res!(is_not!("\r\n"), from_utf8_slice)
-        >> take!(1)
-        >> (text)
-    )
-);
+pub fn m3u_tag(i: &[u8]) -> IResult<&[u8], String> {
+    map_res(tag("#EXTM3U"), from_utf8_slice)(i)
```

Comment:
This will always return the string "#EXTM3U", which seems a bit wasteful. Similar situation in some other places IIRC.

Do you want me to clean that up in a separate commit?

Reply:
I went ahead and removed this part completely, and instead require each playlist to actually start with `#EXTM3U`. This fixes https://github.com/rutgersc/m3u8-rs/issues/27 .

Test is also included.